Use native cmake to precompile headers

gassmoeller opened the pull request:
See the discussion in #3585. We could also enable this unconditionally as long cmake supports it. Or we keep it like this for now and enable it by default later.

tjhei commented on CMakeLists.txt:
```diff
@@ -467,6 +467,12 @@ ENDIF()
 DEAL_II_SETUP_TARGET(${TARGET})
 
+IF (ASPECT_PRECOMPILE_HEADERS)
+  IF(CMAKE_VERSION VERSION_GREATER_EQUAL 3.16)
```
I am not sure I understand. Wouldn't you run into the FATAL_ERROR 6 lines above?

tjhei commented on CMakeLists.txt:
```diff
@@ -467,6 +467,12 @@ ENDIF()
 
+IF (ASPECT_PRECOMPILE_HEADERS)
+  IF(CMAKE_VERSION VERSION_GREATER_EQUAL 3.16)
+    TARGET_PRECOMPILE_HEADERS(aspect PRIVATE <aspect/global.h> <aspect/introspection.h> <aspect/parameters.h> <deal.II/particles/particle_handler.h> <deal.II/base/table_handler.h> <deal.II/base/timer.h> <deal.II/base/conditional_ostream.h> <deal.II/distributed/tria.h> <deal.II/dofs/dof_handler.h> <deal.II/fe/fe.h> <deal.II/fe/mapping_q.h>)
```
is particle_handler useful to have in here?

gassmoeller commented:
You are right, I should test the final version before pushing (it also didnt work with old cmakes, e.g. 3.5 like on the tester), should be fixed now (I tested with cmake 3.5.1, 3.8.2 and 3.16.3).

Having the particle handler header in there saves about 20 s (it is usually included in `simulator_access.h` via `aspect/particle/world.h`, but that one includes all kind of stuff that should not be in the precompiled headers). 
Additionally, excluding the world builder files from including the precompiled headers saves about 15 s.

tjhei commented on CMakeLists.txt:
```diff
@@ -458,6 +460,8 @@ IF (ASPECT_PRECOMPILE_HEADERS)
+  ELSEIF(CMAKE_VERSION VERSION_GREATER 3.15)
+    # Do nothing, this case is handled below
   ELSE()
     MESSAGE(FATAL_ERROR "ASPECT_PRECOMPILE_HEADERS is currently only supported for CMake version 3.8 or less.")
```
can you update this error message please?

tjhei commented on CMakeLists.txt:
```diff
@@ -144,13 +144,15 @@ if(ASPECT_WITH_WORLD_BUILDER)
   # set ASPECT_ENABLE_WORLD_BUILDER define for all source files:
   FOREACH(_source_file ${wb_files})
     SET_PROPERTY(SOURCE ${_source_file} PROPERTY COTIRE_EXCLUDED TRUE ) 
+    SET_PROPERTY(SOURCE ${_source_file} PROPERTY SKIP_PRECOMPILE_HEADERS TRUE )
```
why do we need to skip this? Can you leave a comment?

tjhei commented on CMakeLists.txt:
```diff
@@ -467,6 +471,10 @@ ENDIF()
 DEAL_II_SETUP_TARGET(${TARGET})
 
+IF (ASPECT_PRECOMPILE_HEADERS AND CMAKE_VERSION VERSION_GREATER 3.15)
+  TARGET_PRECOMPILE_HEADERS(aspect PRIVATE <aspect/global.h> <aspect/introspection.h> <aspect/parameters.h> <deal.II/particles/particle_handler.h> <deal.II/base/table_handler.h> <deal.II/base/timer.h> <deal.II/base/conditional_ostream.h> <deal.II/distributed/tria.h> <deal.II/dofs/dof_handler.h> <deal.II/fe/fe.h> <deal.II/fe/mapping_q.h>)
```
would you mind breaking this into one file per line? Maybe add a comment with the reasoning what to include and what to leave out?

gassmoeller commented:
I think I addressed all issues and also move the declaration of the cmake variable to the place where it is used (had to be earlier in the file before we decided to unconditionally reorder helper_functions.cc in #3591). After the results discussed in #3585 I also decided to make precompiling the headers the default for cmake >3.15, do you agree? Ready from my side.

tjhei commented on CMakeLists.txt:
```diff
@@ -513,24 +503,53 @@ CONFIGURE_FILE(
+  SET(ASPECT_PRECOMPILE_HEADERS OFF CACHE BOOL "Precompile external header files using cotire to speedup compile time. Currently only supported for cmake versions older than 3.9 or newer than 3.15.")
+ELSEIF(CMAKE_VERSION VERSION_GREATER 3.15)
+  SET(ASPECT_PRECOMPILE_HEADERS ON CACHE BOOL "Precompile external header files to speedup compile time. Currently only supported for cmake versions older than 3.9 or newer than 3.9.")
+ENDIF()
```
We should declare this variable in all 3 cases. Can you add an else with ``SET(... OFF)``?

tjhei commented on CMakeLists.txt:
```diff
@@ -513,24 +503,53 @@ CONFIGURE_FILE(
-  # and therefore the whole precompilation
-  cotire(aspect)
+  IF (CMAKE_VERSION VERSION_LESS 3.9)
+    MESSAGE(STATUS "Enabling cotire to precompile external header files.")
```
I would suggest we get rid of cotire entirely. It is too complicated to make sure all 3 variants work (and I don't see how we can test this on jenkins). We can delay this until we bump to deal.II 9.2 if you prefer.

tjhei commented:
> decided to make precompiling the headers the default

sounds good.

We need to (separately from this PR) figure out how we want to test using Jenkins. Maybe disable precompiled headers there?

gassmoeller commented:
> I would suggest we get rid of cotire entirely. It is too complicated to make sure all 3 variants work (and I don't see how we can test this on jenkins). We can delay this until we bump to deal.II 9.2 if you prefer.

I agree, but lets do that as follow-up to #3583, when we update the testers. At the moment it is still using cotire.

> We need to (separately from this PR) figure out how we want to test using Jenkins. Maybe disable precompiled headers there?

Currently we run a precompiled header build and a unity build. I would suggest afterwards we run one unity build with precompiled headers (which hopefully catches most/all errors for unity builds and precompiled headers individually), and one build without both. We could split this into your tester vs CIG testers? Your tester is typically faster, so maybe you compile without unity/precompile? For a full test we would need to check all 4 variations.

tjhei commented:
> At the moment it is still using cotire.

okay, I updated  #3583 to use the new unity build because cotire won't work there (newer cmake).

We can delete cotire later.
